sycl/source/detail/graph_impl.cpp

```diff
@@ -1095,6 +1154,246 @@ void exec_graph_impl::duplicateNodes() {
   MNodeStorage.insert(MNodeStorage.begin(), NewNodes.begin(), NewNodes.end());
 }
 
+void exec_graph_impl::update(std::shared_ptr<node_impl> Node) {
+  this->update(std::vector<std::shared_ptr<node_impl>>{Node});
+}
+void exec_graph_impl::update(
+    const std::vector<std::shared_ptr<node_impl>> Nodes) {
+
+  if (!MIsUpdatable) {
+    throw sycl::exception(sycl::make_error_code(errc::invalid),
+                          "update() cannot be called on a executable graph "
+                          "which was not created with property::updatable");
+  }
+
+  // If there are any accessor requirements, we have to update through the
+  // scheduler to ensure that any allocations have taken place before trying to
+  // update.
+  bool NeedScheduledUpdate = false;
+  for (auto &Node : Nodes) {
+    // Check if node(s) derived from this modifiable node exists in this graph
+    if (MIDCache.count(Node->getID()) == 0) {
+      throw sycl::exception(
+          sycl::make_error_code(errc::invalid),
+          "Node passed to update() is not part of the graph.");
+    }
+    if (Node->MCGType != sycl::detail::CG::Kernel) {
+      throw sycl::exception(errc::invalid, "Cannot update non-kernel nodes");
+    }
+
+    if (Node->MCommandGroup->getRequirements().size() == 0) {
+      continue;
+    }
+    NeedScheduledUpdate = true;
+    // Update cached requirements for this graph with updated node ones
+    auto UpdatedReqs = Node->MUpdatedAccessorsCache;
+    for (auto &CachedReq : MRequirements) {
+      for (auto &UpdatedReq : UpdatedReqs) {
+        if (CachedReq == UpdatedReq.first) {
+          CachedReq = UpdatedReq.second;
+        }
+      }
+    }
```

Comment:
It could be a map but thinking about the function of this some more I think it's not actually a sufficient solution. It falls apart if accessors are updated more than once before the graph itself is updated. Since a node accessor could potentially be updated any number of times before this code happens, it's probably not feasible to track all those changes. Probably just rebuilding the list of requirements from scratch is safer, even if it's slightly less performant.